lib/cyclone-server/src/state.rs

```diff
@@ -1,38 +1,59 @@
 use std::{
     ops::Deref,
     path::{Path, PathBuf},
+    process::Stdio,
     sync::Arc,
     time::Duration,
 };
 
 use axum::extract::FromRef;
-use tokio::sync::mpsc;
+use telemetry::tracing::debug;
+use tokio::{
+    process::{Child, Command},
+    sync::{mpsc, Mutex},
+};
+
+use crate::execution::ExecutionError;
+type Result<T> = std::result::Result<T, ExecutionError>;
 
 #[derive(Clone, FromRef)]
 pub struct AppState {
-    lang_server_path: LangServerPath,
-    telemetry_level: TelemetryLevel,
-    lang_server_function_timeout: LangServerFunctionTimeout,
+    child: LangServerChild,
     lang_server_process_timeout: LangServerProcessTimeout,
+    telemetry_level: TelemetryLevel,
 }
 
 impl AppState {
-    pub fn new(
-        lang_server_path: impl Into<PathBuf>,
+    pub async fn new(
+        lang_server_path: impl Into<PathBuf> + std::convert::AsRef<std::ffi::OsStr>,
         telemetry_level: Box<dyn telemetry::TelemetryLevel>,
         lang_server_function_timeout: Option<usize>,
         lang_server_process_timeout: Option<u64>,
-    ) -> Self {
-        Self {
-            lang_server_path: LangServerPath(Arc::new(lang_server_path.into())),
-            telemetry_level: TelemetryLevel(Arc::new(telemetry_level)),
-            lang_server_function_timeout: LangServerFunctionTimeout(Arc::new(
-                lang_server_function_timeout,
-            )),
+    ) -> Result<Self> {
+        let mut cmd = Command::new(&lang_server_path);
+        cmd.stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped());
+
+        if let Some(timeout) = lang_server_function_timeout {
+            cmd.arg("--timeout").arg(timeout.to_string());
+        }
+        if telemetry_level.is_debug_or_lower().await {
+            cmd.env("SI_LANG_JS_LOG", "*");
+        }
+
+        debug!(cmd = ?cmd, "spawning child process");
+        let child = cmd
+            .spawn()
+            .map_err(|err| ExecutionError::ChildSpawn(err, lang_server_path.into()))?;
+
+        Ok(Self {
+            child: LangServerChild(Arc::new(Mutex::new(child))),
```

Comment:
I was wonder why you need all this, then remembered that in Axum the `AppState` must impl `Clone` for each request, and so this make sense. It's probably only really used for that single Cyclone request anyway. Neato

Reply:
Yes, exactly. I tried with just an Arc, but found ~14 strong references to it when I went to unwrap it. I assumed axum was doing this under the covers.